### PR TITLE
Add dtype field to RasterDataset

### PR DESCRIPTION
**Pull request by calebrob6**
In the great DataModule overhaul of winter 2022/2023 (https://github.com/microsoft/torchgeo/pull/992) we made sure that all the dtypes returned by our datasets played nicely with Kornia. As part of this, we made the assumption in RasterDataset that all "mask" layers should be `torch.long` and all images should be `torch.float`. As a result, our RasterDatasets can essentially not be used for real valued regression tasks.

This PR aims to fix this by introducing a new field to RasterDataset called `dtype` that, if set, forces a cast to that type right before the dataset level transform is called. With this we can fix the dtype expectations on a per dataset level.

### CODE REVIEW

File: torchgeo/datasets/geo.py [hunk 1]
```diff
@@ -427,10 +430,14 @@ def __getitem__(self, query: BoundingBox) -> Dict[str, Any]:
             data = self._merge_files(filepaths, query, self.band_indexes)
 
         sample = {"crs": self.crs, "bbox": query}
+
+        if self.dtype is not None:
+            data = data.to(self.dtype)
+
         if self.is_image:
             sample["image"] = data.float()
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:438]: We should remove this cast too for consistency.

**Review by adamjstewart — COMMENTED** [review]
Technically this is an API change (we introduce a new class attribute) so it should wait until 0.5.0, but it also fixes an unintended consequence of #992, so it could also be argued that it should go in 0.4.1. Not sure which to choose.

We'll need to take this PR into account when thinking about #985.

File: torchgeo/datasets/geo.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/datasets/geo.py:438]: all images should be float though

Comment (calebrob6) [reply torchgeo/datasets/geo.py]: let's see what happens though :)

Comment (calebrob6) [reply torchgeo/datasets/geo.py]: It looks like this assumption is built into RasterDataset -- https://github.com/microsoft/torchgeo/blob/main/tests/datasets/test_geo.py#L201

There is a test that has a custom dataset with uint32 dtype that checks to make sure it has been converted by RasterDataset to float

Comment (adamjstewart) [reply torchgeo/datasets/geo.py]: We can remove that test if we don't plan on unifying dtypes across datasets.

**Comment by adamjstewart** [comment]
Just to complicate things, different bands of the same image may have different dtypes: https://github.com/microsoft/torchgeo/discussions/1182

Not sure how to help with this one. We can't stack all bands of the image into one tensor unless the tensor has a single shared dtype.

**Comment by calebrob6** [comment]
> Just to complicate things, different bands of the same image may have different dtypes: #1182

That needs to be handled by casting in a pre-processing step. It doesn't make sense for RasterDataset to have different dtypes for different bands because the input to a neural network needs to be a float32, 16, 8, .. not a int/float mashup.

File: torchgeo/datasets/geo.py (hunk 1 above)
Comment (calebrob6) [reply torchgeo/datasets/geo.py]: I think it makes sense to cast images to floats because that is what will be passed to a network

**Comment by adamjstewart** [comment]
In the case of #1182 it may make more sense for the PIXEL_QA band to be read as a mask instead of an image.

File: torchgeo/datasets/geo.py (hunk 1 above)
Comment (adamjstewart) [reply torchgeo/datasets/geo.py]: And I thought it would make sense to cast masks to long because that is what will be passed to torchmetrics 😆 

I think if a user specifies `dtype=torch.float16` when constructing the dataset and we instead use `torch.float32` just because it's an image, they'll be quite confused and will open a bug report. We should be consistent.

**Comment by adamjstewart** [comment]
The only thing holding this PR back is that `dtype` affects masks but not images. This is very counterintuitive, isn't documented, and there's no warning letting you know this. Is there any reason `dtype` shouldn't affect both images and masks?

**Comment by calebrob6** [comment]
`Conv2d` expects inputs to be float32s, therefore it makes a lot of sense to always make inputs float32s. If the naming is confusing then we can rename it `mask_dtype` or something.

> Is there any reason dtype shouldn't affect both images and masks?

Often the masks will be longs while the inputs will be float32s.

You can try running this:
```
import torch
import torch.nn as nn

conv = nn.Conv2d(3, 3, 3, bias=False)

for dtype in [
    torch.float32,
    torch.float64,
    torch.complex64,
    torch.complex128,
    torch.float16,
    torch.bfloat16,
    torch.uint8,
    torch.int8,
    torch.int16,
    torch.int32,
    torch.int64,
    torch.bool,
]:
    x = torch.ones(3, 9, 9, dtype=dtype)
    try:
        y = conv(x)
        print(f"{dtype} works")
    except:
        print(f"{dtype} doesn't work")
```

You should get:
```
torch.float32 works
torch.float64 doesn't work
torch.complex64 doesn't work
torch.complex128 doesn't work
torch.float16 doesn't work
torch.bfloat16 doesn't work
torch.uint8 doesn't work
torch.int8 doesn't work
torch.int16 doesn't work
torch.int32 doesn't work
torch.int64 doesn't work
torch.bool doesn't work
```

**Comment by adamjstewart** [comment]
Kinda surprised float64 doesn't work.

I would be fine with calling it `mask_dtype`. We can always rename it later if someone provides a use case for changing image dtype.

**Comment by calebrob6** [comment]
Actually, calling it `mask_dtype` doesn't make sense -- RasterDatasets are either images or masks, they only have one dtype.

**Comment by adamjstewart** [comment]
Let's call it dtype then. I would either let it apply to both images and masks, or raise an error when someone uses not-float32 for an image.

**Comment by calebrob6** [comment]
How about a UserWarning? Raising an all-out error seems extreme to me when everything should otherwise work.

**Comment by calebrob6** [comment]
This should help with https://github.com/microsoft/torchgeo/issues/849 btw

File: torchgeo/datasets/geo.py [hunk 2]
```diff
@@ -427,10 +431,19 @@ def __getitem__(self, query: BoundingBox) -> Dict[str, Any]:
...
 
         sample = {"crs": self.crs, "bbox": query}
+
+        if self.dtype is not None and not self.is_image:
+            data = data.to(self.dtype)
+        if self.dtype is not None and self.is_image:
+            warn(
+                "Custom dtype is explicitely set, however the current RasterDataset is"
+                + " an image, so no action will be taken."
+            )
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:441]:
```suggestion
        if self.dtype is not None:
            if self.is_image:
                warn("Custom dtype is explicitly set, but dtype is only valid for mask RasterDatasets.")
            else:
                data = data.to(self.dtype)
```
This feels cleaner

**Comment by adamjstewart** [comment]
Needs tests, see https://docs.pytest.org/en/7.1.x/how-to/capture-warnings.html

File: torchgeo/datasets/chesapeake.py [hunk 3]
```diff
@@ -40,6 +40,7 @@ class Chesapeake(RasterDataset, abc.ABC):
     """
 
     is_image = False
+    dtype = torch.long
```
Comment (adamjstewart) [inline torchgeo/datasets/chesapeake.py:43]: I'm guessing we need to set this for all other current mask datasets?
```console
$ grep 'is_image = False' *.py
agb_live_woody_density.py:    is_image = False
astergdem.py:    is_image = False
cdl.py:    is_image = False
chesapeake.py:    is_image = False
cms_mangrove_canopy.py:    is_image = False
esri2020.py:    is_image = False
eudem.py:    is_image = False
globbiomass.py:    is_image = False
nlcd.py:    is_image = False
```

File: torchgeo/datasets/geo.py [hunk 4]
```diff
@@ -429,10 +433,20 @@ def __getitem__(self, query: BoundingBox) -> dict[str, Any]:
             data = self._merge_files(filepaths, query, self.band_indexes)
 
         sample = {"crs": self.crs, "bbox": query}
+
+        if self.dtype is not None:
+            if self.is_image:
+                warn(
+                    "Custom dtype is explicitly set, but dtype is only valid for mask"
+                    + " RasterDatasets."
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:441]:
```suggestion
                    " RasterDatasets."
```
Fun fact: Python strings automatically concatenate even without the +.

**Review by adamjstewart — COMMENTED** [review]
I'm fine with this approach. We could also use something like:
```python
@property
def dtype(self) -> torch.dtype:
    if self.is_image:
        return torch.float32
    else:
        return torch.long
```
Then we would only need to override this for the small number of datasets that have a different dtype.

File: torchgeo/datasets/geo.py (hunk 4 above)
Comment (calebrob6) [reply torchgeo/datasets/geo.py:441]: Yeah, I'm not a fan lol, if you forget a comma in your list of strings it will happily fail :)

File: torchgeo/datasets/chesapeake.py (hunk 3 above)
Comment (calebrob6) [reply torchgeo/datasets/chesapeake.py:43]: only datasets where the dtype of the mask needs to be overridden

Comment (calebrob6) [reply torchgeo/datasets/chesapeake.py:43]: > I'm fine with this approach. We could also use something like:
> 
> ```python
> @property
> def dtype(self) -> torch.dtype:
>     if self.is_image:
>         return torch.float32
>     else:
>         return torch.long
> ```
> 
> Then we would only need to override this for the small number of datasets that have a different dtype.

But the whole point is that we sometimes want to allow for non `torch.long` masks

Comment (adamjstewart) [reply torchgeo/datasets/chesapeake.py:43]: Yes, and adding `dtype = torch.float32` to a subclass will override this property. If `dtype` isn't set in a subclass, the property will be used instead.

Comment (calebrob6) [reply torchgeo/datasets/chesapeake.py:43]: And then always cast to dtype in `__getitem__(...)`? Why is that better than what we're currently doing in this PR? 

Also, we could set this for the other mask datasets if we wanted to be explicit! Let me know.

Comment (adamjstewart) [reply torchgeo/datasets/chesapeake.py:43]: Many raster files are stored in uint8–uint32. Instead of adding `dtype = torch.long` or `dtype = torch.float32` to all of those datasets, we can set it once in the base class and override it only when a dataset needs to be different (none of our current datasets, but future pixel-wise regression datasets).

Comment (adamjstewart) [inline torchgeo/datasets/chesapeake.py:43]: This can be removed now

File: torchgeo/datasets/geo.py [hunk 5]
```diff
@@ -296,6 +296,18 @@ class RasterDataset(GeoDataset):
     #: Color map for the dataset, used for plotting
     cmap: dict[int, tuple[int, int, int, int]] = {}
 
+    @property
+    def dtype(self) -> torch.dtype:
+        """The dtype of the dataset (overrides the dtype of the data file via a cast).
+
+        Returns:
+            the dtype of the dataset
```
Comment (adamjstewart) [inline torchgeo/datasets/geo.py:304]: Can you add versionadded?
